### ISSUE TITLE
how to model multiple associations to one table

### ISSUE THREAD

**Issue by barbalex** [Nozbe/WatermelonDB#885]
In my app plants are delivered to nurseries.

Deliveries can originate from nurseries or terminate there.
Thus the deliveries table has two relations to the nurseries table with the corresponding fields being: 
- `from_nursery_id` 
- `to_nursery_id`

I can create an association for one of these in the delivery table:
```js
static associations = {
  nursery: { type: 'belongs_to', key: 'from_nursery_id' }
}
```

But how can I build the second?
An object cannot have two keys with the same name...

I had originally created them using other names:
```js
static associations = {
  from_nursery: { type: 'belongs_to', key: 'from_nursery_id' },
  to_nursery: { type: 'belongs_to', key: 'to_nursery_id' }
}
```

Until watermelondb told me there were missing associations. Then I realized the key has to be the table name.

The same problem occurs at the other side of the relations, at the nursery table. There I originally had:
```js
static associations = {
  from_delivery: { type: 'has_many', foreignKey: 'from_nursery_id' },
  to_delivery: { type: 'has_many', foreignKey: 'to_nursery_id' }
}
```
and it seems I can only add one of the two associations here too.

What will be the downsides of missing associations?

**Comment by barbalex** [comment]
I just realized I can't query nursery from delivery using `Q.on` for both relations because that probably looks up the delivery side field in it's associations. So I can only query the one relation.

**Comment by barbalex** [comment]
I am currently circumventing the problem by querying and filtering the data  where it is needed instead of using references. I suppose that modeling multiple associations is not possible with the current implementation in WatermelonDB and will close this issue to make room for others.

**Comment by CDBridger** [comment]
I don't think this issue should be closed either just because it was circumvented

**Comment by KrisLau** [comment]
@CDBridger I think it might not be possible with the current implementation of WDB which is why it was closed

**Comment by KrisLau** [comment]
@radex Would this be possibly added as a future feature? I think ideally it would be nice if in the model (using OP example), we are able to do something like: 
```
  nursery: { type: 'belongs_to', key: ['from_nursery_id', 'to_nursery_id'] }
```

**Comment by Rodrigobanselmo** [comment]
any updates on this? same problem here!

**Comment by PEZO19** [comment]
@radex @KrisLau @Rodrigobanselmo @g-nogueira @CDBridger @barbalex 

Haven't try that yet, but maybe works. Came up with the idea myself, but maybe there is a widely known SQL pattern related to that, it feels to me there must be a "named pattern" for that. Feels similar to polymorphic types in SQL a bit, but a bit different, because here we have "edge types" instead "table types".

When 2 entities (here: Nursery, Delivery) are "connected" with a pivot / relation, it is also possible, that they are "connected" through "multiple kind of edges" (here: "from edge"/"to edge", instead just "plain edge"). In that case, we might store this "edge type"/"edge kind" in an extra column on the relation/"pivot" table. (I prefer the word "kind" here, as type is overloaded in general.)

"Type/Kind-Extended" Relation/pivot table:
| nursery_id | delivery_id | edge_kind |
|------------|-------------|-----------|
| 1          | 1           | "from"    |
| 2          | 1           | "to"      |

Given that, you can use that extra `edge_kind` information in your `Nursery` / `Delivery` models to filter the proper records.

If I am missing something and/or you know the related SQL patterns name, please let me know.

**Comment by kanounzied** [comment]
@radex  is this still not solved yet ? are there any news ? In my case I am using a simple logging friendly app with many fields created_by and updated_by and deleted_by all pointing to the same table users.
